extensions/src/main/java/org/pentaho/platform/plugin/services/email/EmailConfiguration.java

```diff
@@ -50,14 +50,27 @@ public class EmailConfiguration implements Serializable, IEmailConfiguration {
   private String password;
   private boolean useSsl;
   private boolean useStartTls;
+  private String authMechanism;
+  private String clientId;
+  private String clientSecret;
+  private String tokenUrl;
+  private String scope;
+
+  private String grantType;
+  private String refreshToken;
+
+  private String authorizationCode;
+  private String redirectUri;
 
   public EmailConfiguration() {
   }
 
   public EmailConfiguration( final boolean authenticate, final boolean debug, final String defaultFrom,
       final String fromName, final String smtpHost, final Integer smtpPort, final String smtpProtocol,
       final boolean smtpQuitWait, final String userId, final String password, final boolean useSsl,
-      final boolean useStartTls ) {
+      final boolean useStartTls, final String authMechanism, final String clientId, final String clientSecret,
```

Comment:
We are now at 21 arguments for a constructor, it is now unmanageable. Is the 20th argument String for refreshToken,  authorizationCode!?

Implement builder pattern: https://www.vogella.com/tutorials/DesignPatternBuilder/article.html

it is normally to have a single constructor for the builder and those arguments are "required".  But since, basically all are required for MS Graph api Email Configuration should have it's own builder and possibly extends EmailConfigurationBuilder.

Reply:
I agree with using the builder pattern. For immutable classes, receiving the builder itself as a constructor parameter avoids having a constructor with unbounded parameters.

However, in this case, given that the `IEmailConfiguration` interface is not immutable (has set methods), it may be acceptable that the new configuration properties are not made available as constructor parameters, and that it would be required to call the corresponding set methods to set them (similar API experience to that of the builder).

Reply:
Good point. I agree builders should only be used immutable objects. Then, I don't see much value in maintaining a constructor with huge number of arguments. 

In my opinion, I would keep the constructor with a few arguments as possible and then just call the new setters. I don't see the constructor with arguments being used in production anyway.
Leave it up to the author.

Reply:
Already addressed.